### PR TITLE
Feature Enhancements, Bug Fixes, and Updates for Improved Functionality and User Experience

### PR DESCRIPTION
**Pull request by CarlosZiegler**
# Pull Request Description

This pull request introduces several new features, bug fixes, and updates to improve the functionality and user experience of the application.

## New Features

- Added validation to POST request body in `api/chat/route.ts`.
- Introduced Google login button in `sign-in/page.tsx`.
- Added Google provider to NextAuth in `auth.ts`.
- Included model selection in the chat component `chat.tsx`.
- Added model selector to the header in `header.tsx`.
- Created a new `ModelSelector` component in `model-selector.tsx`.
- Added a popover for template selection in `prompt-form.tsx`.
- Introduced `Command` component and related sub-components in `ui/command.tsx`.
- Added `HoverCard`, `Popover` components in `ui`.
- Introduced `IconGoogle` component in `ui`.
- Added GPT-4 models and types in `models.ts`.
- Included support for model compatibility types in `models.ts`.
- Added `openAImodels` array and `models` array in `models.ts`.
- Introduced `modelAtom` to store the selected model in `store.ts`.
- Added `zValidateReq` function to validate environment variables in `validate.ts`.
- Added `@radix-ui/react-hover-card` and `@radix-ui/react-popover` dependencies in `package.json`.
- Included `jotai`, `lucide-react`, and `zod` dependencies in `package.json`.

## Bug Fixes

- Fixed issue in `actions.ts` to only execute pipeline if there are chats to clear.

## Updates

- Added environment variables schema and validation in `envs.ts`.

These changes aim to enhance the overall functionality and user experience of the application.

Screenshots:

- Action button: 

image

- Model selector:
image

-Google Login
image

### CODE REVIEW

**Comment by CarlosZiegler** [comment]
I added accidentally more commit, while test deploy :( But reverted some changes.

File: components/ui/command.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
+"use client"
+
+import * as React from "react"
+import { DialogProps } from "@radix-ui/react-dialog"
+import { Command as CommandPrimitive } from "cmdk"
+import { Search } from "lucide-react"
```
Comment (jaredpalmer) [inline components/ui/command.tsx:6]: Remove `lucide-react`, use search icon from heroicons and place into icons file. (aside: unoptimized lucide is terrible for performance)

File: components/ui/icons.tsx [hunk 2]
```diff
@@ -482,6 +512,7 @@ export {
   IconOpenAI,
   IconVercel,
   IconGitHub,
+  IconGoogle,
   IconSeparator,
   IconArrowDown,
   IconArrowRight,
```
Comment (jaredpalmer) [inline components/ui/icons.tsx:518]: Remove google icon

File: constants/envs.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  KV_REST_API_READ_ONLY_TOKEN: z.string()
+})
+
+const _env = envsSchema.safeParse(process.env)
+
+if (!_env.success) {
+  console.error('❌ Invalid environment variables:\n')
+  _env.error.issues.forEach(issue => {
+    console.error(issue)
+  })
```
Comment (jaredpalmer) [inline constants/envs.ts:19]: Let's remove so we stay unopinionated about these kinds of things

File: constants/models.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    name: 'babbage',
+    description: 'Capable of straightforward tasks, very fast, and lower cost.',
+    type: 'GPT-3',
+    compatibility: 'fine-tunes',
+    tokens: '2,049 tokens',
+    strengths: 'Straightforward tasks, speed, cost-effectiveness'
+  },
+  {
+    id: 'ada',
+    name: 'ada',
```
Comment (jaredpalmer) [inline constants/models.ts:195]: Lets remove non-chat models. Folks can use sdk.vercel.ai to play with these in a proper prompt experience.

File: constants/models.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    description:
+      'Capable of very simple tasks, usually the fastest model in the GPT-3 series, and lowest cost.',
+    type: 'GPT-3',
+    compatibility: 'fine-tunes',
+    tokens: '2,049 tokens',
```
Comment (jaredpalmer) [inline constants/models.ts:200]: This heavily restricts the starter template to OpenAI as it is implemented now. Let's split this into another PR and work through the DX of multiple providers and multiple models more thoroughly.

File: lib/store.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import { models, Model } from '@/constants/models'
+import { atom } from 'jotai'
```
Comment (jaredpalmer) [inline lib/store.ts:2]: As mentioned, lets drop this for plain React and props

File: lib/validate.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { AnyZodObject, z } from 'zod'
+
+export async function zValidateReq<T extends AnyZodObject>(
+  schema: T,
+  req: Request
+): Promise<z.infer<T>> {
+  const json = await req.json()
+  const input = schema.safeParse(json)
+  if (!input.success) {
+    console.error('❌ Invalid environment variables:\n')
```
Comment (jaredpalmer) [inline lib/validate.ts:10]: Incorrect error messages.

**Review by jaredpalmer — CHANGES_REQUESTED** [review]
Epic PR! Thanks for this.

File: components/chat.tsx [hunk 8]
```diff
@@ -19,6 +19,8 @@ import {
 import { useState } from 'react'
 import { Button } from './ui/button'
 import { Input } from './ui/input'
+import { useAtom } from 'jotai'
```
[the remark this answers is not in the document]
Comment (CarlosZiegler) [reply components/chat.tsx:22]: OK, I understand, I am new on the new version of Next and App routers,  what is the best way to share state in App routes version ? Can you share a sinppet ? The I will refactor :)

File: components/ui/command.tsx (hunk 1 above)
Comment (CarlosZiegler) [reply components/ui/command.tsx:6]: OK I used this shadcn cli to create component, I will refactor this :)

File: components/chat.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaredpalmer) [reply components/chat.tsx:22]: Use `useState` and pass down data/functions as a props

File: .env.example [hunk 9]
```diff
@@ -11,9 +11,12 @@ NEXTAUTH_URL=http://localhost:3000
 AUTH_GITHUB_ID=XXXXXXXX
 AUTH_GITHUB_SECRET=XXXXXXXX
 
-# instructions to create kv database here: https://vercel.com/docs/storage/vercel-kv/quickstart and 
+## Create a Google OAuth app here: https://developers.google.com/identity/protocols/oauth2
+AUTH_GOOGLE_CLIENT_ID=521329335604-j4cnrcs5ng2i6070noq2ku6q8gon0hol.apps.googleusercontent.com
```
Comment (jaredpalmer) [inline .env.example:15]: Remove these since we removed google auth

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

File: components/chat.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (CarlosZiegler) [reply components/chat.tsx:22]: OK props drilling, but question i  where. In layout ? then I need to move metaData to another place or ?

[the remark this answers is not in the document]
Comment (CarlosZiegler) [reply components/chat.tsx]: First approach, move select to promp:

image

image
